clippy_lints/src/methods/inspect_then_for_each.rs

```diff
@@ -0,0 +1,20 @@
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+
+use crate::utils::{match_trait_method, paths, span_lint_and_help};
+
+use super::INSPECT_THEN_FOR_EACH;
+
+/// lint use of `inspect().for_each()` for `Iterators`
+pub(super) fn lint<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'_>,
+    _inspect_args: &'tcx [hir::Expr<'_>],
+    _for_each_args: &'tcx [hir::Expr<'_>],
+) {
+    if match_trait_method(cx, expr, &paths::ITERATOR) {
+        let msg = "called `inspect(..).for_each(..)` on an `Iterator`";
+        let hint = "this is more succinctly expressed by calling ` for_each(..)` instead";
```

Comment:
```suggestion
        let hint = "move the code from `inspect(..)` to `for_each(..)` and remove the `inspect(..)`";
```

I think a `span_lint_and_help` is enough here for now. Building a good looking suggestion by combining the closures into one is way more involved (getting the indentation right and whatnot).

---

Some guidance how this can be done:

1. extract the closure bodies from `_inspect_args[1]` and ``_for_each_args[1]``
2. check if the closure body of ``for_each`` is already a block
    a. If it is, get the span of everything inside the block.
    b. If not, just get the span of the expression
3. build the suggestion with a block around it, where the body of `inspect` is placed at the top and is indented the same as everything inside the `for_each` function.

This will involve much fiddling. I done this before and utils has many methods to achieve this (search for usages of ``indent_of`` in the code base). It'll be mostly trial and error though.

Anyway, this should probably be done in a separate PR.

Reply:
You're right that it could be done in a separate PR. In any case, I've taken notes about your suggestions :)